### PR TITLE
Remove disk feature

### PR DESCRIPTION
**Pull request by alex-necula**
| Q               | A      |
| --------------- | ------ |
| Bug fix?        | no |
| New feature?    | yes |
| API breaks?     | no |
| Deprecations?   | no |
| Related tickets | internal |
| License         | Apache 2.0 |

### What's in this PR?

This PR adds the disk removal functionality. It is based on Cruise Control's new Disk Removal endpoint: https://github.com/adobe/cruise-control/pull/12

The following logic is followed, in a successful scenario:
- koperator compares the list of existing mount paths with the desired mount paths from the spec
- if a removal is detected, it will mark the disk for removal
- a CC task is created that will drain the data from the removed disks
- KafkaCluster will have "ClusterReconciling" state until the disk is drained successfully, no rolling upgrades will be done until it completes
- if the operation completes, the Rolling Upgrade is performed, PVC is removed

The following logic is followed, in a failure scenario:
- cluster is stuck in ClusterReconciling state because the disk removal failed
- the user should put back all the disks he removed (completely rollback the disk removal, otherwise the cluster will still be stuck in ClusterReconciling, since CC returns error for the whole disk removal operation)
- the readded disks will be marked for rebalance and a disk rebalance will be performed
- no rolling upgrades will be done

### Why?

Previously, disk removal was blocked because there was a risk of data loss. Therefore, disk downscale was not possible, incurring additional costs for unused disks, that were upscaled in the past.

### Additional context

This PR was tested locally using kind. 

During the time between the disk removal starts and the moment the rolling upgrade is performed, Kafka may assign new topics to the removed disks. This presents a potential data loss risk, but I am not aware of methods that would mitigate this risk. During the downscale, users should be informed not to create new topics.

cc @ilievladiulian 

### Checklist
- [x] Implementation tested
- [x] Error handling code meets the [guideline](https://github.com/banzaicloud/developer-guide/blob/master/docs/coding-style/error-handling-guide.md)
- [x] Logging code meets the guideline
- [x] User guide and development docs updated (if needed)

### To Do
- [ ] update go-cruise-control version with new release: https://github.com/adobe/go-cruise-control/pull/1

### CODE REVIEW

File: pkg/scale/scale.go [hunk 1]
```diff
@@ -516,6 +521,55 @@ func (cc *cruiseControlScaler) RebalanceWithParams(params map[string]string) (*R
 	}, nil
 }
 
+func (cc *cruiseControlScaler) RemoveDisksWithParams(params map[string]string) (*Result, error) {
```
Comment (alex-necula) [inline pkg/scale/scale.go:524]: This part of code will be changed once go-cruise-control PR is released

File: pkg/scale/scale.go [hunk 2]
```diff
@@ -50,21 +51,25 @@ const (
...
-		paramRebalanceDisk:  {},
-		paramExcludeDemoted: {},
-		paramExcludeRemoved: {},
-	}
+		ParamDestbrokerIDs:  {},
+		ParamRebalanceDisk:  {},
+		ParamExcludeDemoted: {},
+		ParamExcludeRemoved: {},
+	}
+	// TODO use this map to validate the parameters
```
Comment (alex-necula) [inline pkg/scale/scale.go:69]: This part of code will be changed once go-cruise-control PR is released

**Comment by alex-necula** [comment]
PR waiting for successful upstream rebase

**Comment by amuraru** [comment]
> PR waiting for successful upstream rebase


what do you mean?

**Comment by alex-necula** [comment]
> adobe/cruise-control#7

@aguzovatii performed a rebase on master branch with the latest upstream changes but it was not successful. Waiting for him to fix the conflicts. Until then, this PR has the backup branch as base

**Comment by alex-necula** [comment]
Performed a rebase using the latest upstream changes.

Tests will pass once #54 is merged

**Comment by alex-necula** [comment]
PR is fully tested with https://github.com/adobe/cruise-control/pull/12 and ready for review.
The only thing that has to be changed is the local replacement once the adobe fork is created

**Review by dobrerazvan — APPROVED** [review]

**Review by cristianpetrache — APPROVED** [review]

File: go.mod [hunk 3]
```diff
@@ -140,6 +139,7 @@ require (
 )
 
 replace (
+	github.com/banzaicloud/go-cruise-control => github.com/alex-necula/go-cruise-control v0.6.0-rc5 // TODO this will be replaced with adobe fork once it's created
```
Comment (alex-necula) [inline go.mod:142]: Update with new release once https://github.com/adobe/go-cruise-control/pull/1 is merged

**Pull request merged**